File: ros1bookcode-master-rw/chapt5/chapt5_ws/src/demo_cpp_tf/src/tf_listener.cpp
#include <ros/ros.h>
#include <tf2/utils.h>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>
#include <geometry_msgs/TransformStamped.h>

class TFListener {
 public:
  TFListener() : buffer_(), listener_(buffer_) {
    timer_ = nh_.createTimer(ros::Duration(5.0), &TFListener::getTransform, this);
  }

  void getTransform(const ros::TimerEvent&) {
    try {
      const auto transform = buffer_.lookupTransform("base_link", "target_point", ros::Time(0), ros::Duration(1.0));
      const auto &translation = transform.transform.translation;
      const auto &rotation = transform.transform.rotation;
      double yaw, pitch, roll;
      tf2::Quaternion q(rotation.x, rotation.y, rotation.z, rotation.w);
      tf2::Matrix3x3(q).getEulerYPR(yaw, pitch, roll);
      ROS_INFO("平移分量: (%f, %f, %f)", translation.x, translation.y, translation.z);
      ROS_INFO("旋转分量: (%f, %f, %f)", roll, pitch, yaw);
    } catch (tf2::TransformException &ex) {
      ROS_WARN("异常: %s", ex.what());
    }
  }

 private:
  ros::NodeHandle nh_;
  tf2_ros::Buffer buffer_;
  tf2_ros::TransformListener listener_;
  ros::Timer timer_;
};

int main(int argc, char **argv) {
  ros::init(argc, argv, "tf_listener");
  TFListener node;
  ros::spin();
  return 0;
}

File: ros1bookcode-master-rw/chapt5/chapt5_ws/src/demo_python_tf/demo_python_tf/tf_listener.py
#!/usr/bin/env python3
import rospy
import tf2_ros
from tf.transformations import euler_from_quaternion

class TFListener:
    def __init__(self):
        self.buffer_ = tf2_ros.Buffer()
        self.listener_ = tf2_ros.TransformListener(self.buffer_)
        self.timer_ = rospy.Timer(rospy.Duration(1.0), self.get_transform)

    def get_transform(self, _event):
        try:
            result = self.buffer_.lookup_transform('base_link', 'bottle_link', rospy.Time(0), rospy.Duration(1.0))
            transform = result.transform
            rotation_euler = euler_from_quaternion([
                transform.rotation.x,
                transform.rotation.y,
                transform.rotation.z,
                transform.rotation.w,
            ])
            rospy.loginfo(f'平移:{transform.translation},旋转四元数:{transform.rotation}:旋转欧拉角:{rotation_euler}')
        except Exception as e:
            rospy.logwarn(f'不能够获取坐标变换，原因: {str(e)}')

if __name__ == '__main__':
    rospy.init_node('tf2_listener')
    TFListener()
    rospy.spin()
